src/Features/Core/Portable/ExternalAccess/UnitTesting/UnitTestingFeaturesReferencesService.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeLens;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.UnitTesting;
+
+internal interface IUnitTestingCodeLensContext
+{
+    Task<ImmutableArray<ReferenceMethodDescriptor>?> FindReferenceMethodsAsync(
+        Guid projectGuid, string filePath, TextSpan span, DocumentId? sourceGeneratedDocumentId, CancellationToken cancellationToken);
+}
+
+internal interface IUnitTestingFeaturesReferencesServiceCallback
+{
+    Task<TResult> InvokeAsync<TResult>(string targetName, IReadOnlyList<object?> arguments, CancellationToken cancellationToken);
+}
+
+internal static class UnitTestingFeaturesReferencesService
+{
+    internal static async Task<ImmutableArray<(string MethodFullyQualifedName, string MethodOutputFilePath)>> GetCallerMethodsAsync(
+        Guid projectGuid,
+        string filePath,
+        TextSpan span,
+        DocumentId? sourceGeneratedDocumentId,
```

Comment:
This doesn't seem part of the original contract. Is this needed?

Reply:
Ah I guess you are thinking we would include the code for `GetSourceGeneratorDocumentId` on our side and pass in whatever is returned from that?

I think it may be safer to just pass down the `descriptorContext` directly instead. That way if there is any change to the names of the property bag properties for the guids (like `RoslynDocumentIdGuid`) in the future, we would only need to make the update in one repo (as opposed to in two repos otherwise).

Reply:
The problem is fundamentally that CodeLensDescriptorContext is defined in `Microsoft.VisualStudio.Language.CodeLens` which is precisely what we do not want being referenced :)

Reply:
Ah hmmm ... Got it :)

Reply:
It would still be great to avoid hardcoding the names of the guids on UT side. Could you add a helper in EA to extract the guids from the property bag?

Reply:
done!